### DISCUSSION TITLE
코드의 출력결과와, 그렇게 되는 이유를 선언적 렉시컬환경과 관련지어서 설명해주세요

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#77] [실행컨텍스트]
```js
let x = 1;

if (true){
	let x = 2;
	console.log(x);
}
console.log(x);
```

**Comment by pipisebastian** [comment]
> 코드의 출력결과와, 그렇게 되는 이유를 선언적 렉시컬환경과 관련지어서 설명해주세요

- `2`, `1`
- if 문을 만나게 되면, if 코드 블록의 렉시컬 환경을 새롭게 만들고, 현재 실행중인 실행 컨텍스트는, 해당 렉시컬 환경을 참조합니다.
- if 코드 블럭이 종료된 후에야, 이전에 참조하던 렉시컬 환경(전역)을 다시 참조하게 됩니다.
- if 코드 블럭의 렉시컬 환경은 더이상 참조되고 있지 않기 때문에, 가바지 컬렉터에 의해 추후 메모리가 해제됩니다.

**Comment by publdaze** [comment]
> - 코드의 출력결과와, 그렇게 되는 이유를 선언적 렉시컬환경과 관련지어서 설명해주세요
>   ```js
>   let x = 1;
>   
>   if (true){
> 	  let x = 2;
> 	  console.log(x);
>   }
>   console.log(x);
>   ```

- x가 전역에서 let으로 선언되어 있으므로 해당 식별자는 전역 렉시컬환경의 선언적 렉시컬 환경에 등록되고 관리됨
- 다음 if문 안에 x가 let으로 선언이되어 있는데, let 키워드의 경우 블록 레벨 스코프를 가짐
- 따라서 if문에 해당하는 코드 블럭은 독자적인 렉시컬 환경으로 생성되고 실행중인 실행 컨텍스트는 해당 코드 블럭의 렉시컬 환경을 참조함, 첫번째 출력의 x를 만나면 식별자를 탐색하는데 실행중인 실행 컨텍스트의 렉시컬 환경부터 식별자를 검색하게 되고, 식별자가 바로 존재하므로 할당된 값인 2가 출력됨
- 해당 코드 블럭의 실행이 끝나면 실행 컨텍스트는 다시 전역 렉시컬 환경을 참조하게 되고, 코드 블럭 렉시컬 환경은 더이상 참조되고 있기 않기 때문에 가비지 컬렉터에 의해 추후 메모리 해제됨
- 따라서 마지막 출력부분에서 참조하고 있는 x는 현재 실행중인 실행 컨텍스트가 참조하고 있는 렉시컬 환경이 전역 렉시컬 환경에서 식별자를 검색하게 되고, 해당 식별자에는 1이 할당되어 있으므로 1이 출력됨
